src/iosource/Manager.cc

```diff
@@ -59,124 +109,411 @@ IOSource* Manager::FindSoonest(double* ts)
 			break;
 			}
 
-	// Ideally, we would always call select on the fds to see which
-	// are ready, and return the soonest. Unfortunately, that'd mean
-	// one select-call per packet, which we can't afford in high-volume
-	// environments.  Thus, we call select only every SELECT_FREQUENCY
-	// call (or if all sources report that they are dry).
+	// If there aren't any sources and exit_only_after_terminate is false, just
+	// return an empty set of sources. We want the main loop to end.
+	if ( Size() == 0 && ( ! BifConst::exit_only_after_terminate || terminating ) )
+		return ready;
 
-	++call_count;
+	double timeout = -1;
+	IOSource* timeout_src = nullptr;
+	timer_expired = false;
 
-	IOSource* soonest_src = 0;
-	double soonest_ts = 1e20;
-	double soonest_local_network_time = 1e20;
-	bool all_idle = true;
+	PktSrc* pkt_src = pkt_srcs.empty() ? nullptr : pkt_srcs.front();
 
-	// Find soonest source of those which tell us they have something to
-	// process.
-	for ( SourceList::iterator i = sources.begin(); i != sources.end(); ++i )
+	// If we have an active source of data then find the next timeout value from
+	// those sources. Start by getting the next timeout from the timer manager
```

Comment:
This begs the question: Should the timer mgr become an IOSource?

Reply:
You know what, I actually debated that when I originally did this. I opted to leave it because the `Process()` method wouldn't actually do anything, but maybe it's worth just implementing it that way.

Reply:
> If that's the only reason, I'd say it's worth it, as then we can avoid special-casing it.

The only other good reason is to know whether the timeout was because of the timer manager or because of something else. I'm already planning to remove the `timer_expired` variable and just check for an empty `ready` vector in `net_run`. I wouldn't want to insert the timer manager in that case (or special case checking for it and not inserting it), since it would break that.

Reply:
I cleaned up a bunch of code and converted TimerMgr to an IOSource in https://github.com/zeek/zeek/pull/681/commits/5f6499c5c40360b1511776863beb017c64dcc0a9, if you want to take a look at that specifically. There's a TODO in there that I don't have a great answer for, but it works as is right now. I'll move that if statement somewhere better.